design/handle-backup-of-volumes-by-resources-filters.md

```diff
@@ -0,0 +1,246 @@
+ 
+## Non-Goals
+- We only handle volumes for backup and do not support restore.
+- Currently, only handles volumes, and does not support other resources.
+- Only environment-unrelated and platform-independent general volumes attributes are supported, do not support volumes attributes related to a specific environment.
+ 
+## Use-cases/Scenarios
+### Skip backup volumes by some attributes
+Users want to skip PV with the requirements:
+- option to skip all PV data
+- option to skip specified PV type (RBD, NFS)
+- option to skip specified PV size
+- option to skip specified storage-class
+ 
+## High-Level Design
+Add a new flag `backup-policies-from-file` when executing `velero backup create`, which imports the defined volume resources policies in one YAML file. the YAML file defined volume resources policies for the current backup.
+ 
+When Velero handles volumes backup should respect the policies defined in the imported YAML file.
+ 
+## Detailed Design
+The volume resources policies should contain a list of policies which is the combination of conditions and related `action`, when target volumes meet the conditions, the related `action` will take effection.
+
+Below is the API Design for the user configuration:
+
+### API Design
+```go
+// Action defined one action for a specific way of backup
+type Action struct {
+	// Type defined specific type of action, it could be 'file-system-backup', 'volume-snapshot', or 'skip' currently
+	Type string `yaml:"type,omitempty"`
+	// Parameters defined map of parameters when executing a specific action
+	// +optional
+	// +nullable
+	Parameters map[string]interface{} `yaml:"parameters,omitempty"`
+}
+
+// VolumePolicy defined policy to conditions to match Volumes and related action to handle matched Volumes
+type VolumePolicy struct {
+	// Conditions defined list of conditions to match Volumes
+	Conditions map[string]interface{} `yaml:"conditions,omitempty"`
+	Action     Action                 `yaml:"action,omitempty"`
+}
+
+// ResourcePolicies currently defined slice of volume policies to handle backup
+type ResourcePolicies struct {
+	VolumePolicies []VolumePolicy `yaml:"volumePolicies,omitempty"`
+	// we may support other resource policies in the future, and they could be added separately
+  	// OtherResourcePolicies: []OtherResourcePolicy
+}
+```
+
+The policies YAML config file would look like this:
+```yaml
+version: v1
+volumePolicies:
+# it's a list and if the input item matches the first policy, the latters will be ignored
+# each policy consists of a list of conditions and an action
+
+# each key in the object is one condition, and one policy will apply to resources that meet ALL conditions
+- conditions:
+    # capacity condition matches the volumes whose capacity falls into the range
+    capacity: "0,100Gi"
+    csi:
+      driver: aws.ebs.csi.driver
+      fsType: ext4
+    storageClass:
+    - gp2
+    - ebs-sc
+  action:
+    type: volume-snapshot
```

Comment:
What if I specify action that cannot be completed? Say "volume-snapshot" but I have not created any VolumeSnapshotLocation and CSI feature is disabled. Or "file-system-backup" but velero is installed without restic/kopia? Will the backup partially fail or will the volume be silently skipped?

Reply:
All the policies are responsible for finding matched volumes, whether the  matched volumes could be backup or not is related to velero backup logic, so it will be partially failed based on your scenario